src/domains/heap_domain.cpp

```diff
@@ -1655,3 +1691,354 @@ void heap_domaint::undo_restriction()
+  if(!input_bindings.is_true())
+  {
+    return input_bindings;
+  }
+  return true_exprt();
+}
+
+
+/*******************************************************************\
+
+Function: heap_domaint::clear_pointing_rows
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+void heap_domaint::clear_pointing_rows(
+  const heap_domaint::rowt &row,
+  heap_domaint::heap_valuet &value)
+{
+  heap_domaint::heap_row_valuet &row_value=
+    static_cast<heap_domaint::heap_row_valuet &>(value[row]);
+
+  std::vector<heap_domaint::rowt> to_remove;
+  for(auto &ptr : row_value.pointed_by)
+  {
+    if(ptr!=row)
+    {
+      value[ptr].clear();
+      to_remove.push_back(ptr);
+    }
+  }
+  for(auto &r : to_remove)
+    row_value.pointed_by.erase(r);
+}
+
+/*******************************************************************\
+
+Function: heap_domaint::get_points_to_dest
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: Get an address where the given pointer points to in the current
+          solver iteration. Returns nil_exprt if the value of the pointer
+          is nondet.
+
+\*******************************************************************/
+const exprt heap_domaint::get_points_to_dest(
+  const exprt &solver_value,
+  const exprt &templ_row_expr)
+{
+  // Value from the solver must be converted into an expression
+  exprt ptr_value=value_to_ptr_exprt(solver_value);
+
+  if((ptr_value.id()==ID_constant &&
+      to_constant_expr(ptr_value).get_value()==ID_NULL) ||
+     ptr_value.id()==ID_symbol)
+  {
+    // Add equality p == NULL or p == symbol
+    return ptr_value;
+  }
+  else if(ptr_value.id()==ID_address_of)
+  {
+    // Template row pointer points to the heap (p = &obj)
+    assert(ptr_value.id()==ID_address_of);
+    if(to_address_of_expr(ptr_value).object().id()!=ID_symbol)
+    {
+      // If solver did not return address of a symbol, it is considered
+      // as nondet value.
+      return nil_exprt();
+    }
+
+    symbol_exprt obj=to_symbol_expr(
+      to_address_of_expr(ptr_value).object());
+
+    if(obj.type()!=templ_row_expr.type() &&
+       ns.follow(templ_row_expr.type().subtype())!=ns.follow(obj.type()))
+    {
+      if(!is_cprover_symbol(templ_row_expr))
+      {
+        // If types disagree, it's a nondet (solver assigned random value)
+        return nil_exprt();
+      }
+    }
+
+    // Add equality p == &obj
+    return obj;
+  }
+  else
+    return nil_exprt();
+}
+
+/*******************************************************************\
+
+Function: heap_domaint::find_symbolic_path
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: Find symbolic path that is explored by the current solver
+          iteration. A path is specified by a conjunction of literals
+          containing loop-select guards of all loops in program.
+
+\*******************************************************************/
+void heap_domaint::find_symbolic_path(
```

Comment:
Move this function to `strategy_solver_base`. It willl only accept the `current_guard` parameter, which can be retrieved from the domain (introduce new method for the domain interface: `get_current_loop_guard`).